### PR TITLE
[ValueTracking] isNonZero sub of ptr2int's with recursive GEP

### PR DESCRIPTION
**Pull request by bipmis**
When the sub arguments are ptr2int it is not possible to determine computeKnownBits() of its arguments.
For scalar case generally sub of 2 ptr2int are converted to sub of indexes.
However a loop with recursive GEP/PHI where the arguments to sub is of type ptr2int, if it is possible to determine that a sub of this GEP and another pointer with the same base is KnownNonZero we can return this. This helps subsequent passes to optimize the loop further.

### CODE REVIEW

**Comment by bipmis** [comment]
Gentle Ping

**Comment by bipmis** [comment]
Ping

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -2410,9 +2410,76 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
...
+  // sub(PtrtoInt(LHS), PtrtoInt(RHS))
+  if (match(X, m_PtrToInt(m_Value(LHS))) &&
+      match(Y, m_PtrToInt(m_Value(RHS)))) {
+    GEPOperator *LHSGEP = dyn_cast<GEPOperator>(LHS);
+    // Make sure LHS GEP pointer operand is a PHI. If sub(A,B) is non zero, so
+    // is sub(B,A)
+    if (!LHSGEP || !isa<PHINode>(LHSGEP->getPointerOperand())) {
+      LHSGEP = dyn_cast<GEPOperator>(RHS);
+      std::swap(LHS, RHS);
+    }
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:2432]: Imo variables should by `X` / `Y` instead of `LHS`/`RHS` they get swapped around.
The `LHS`/`RHS` variable names make it seem like their side matters.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -2410,9 +2410,76 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
...
+    if (LHSGEP && isa<PHINode>(LHSGEP->getPointerOperand())) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:2433]: imo early return if either of these are not true.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -2410,9 +2410,76 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
...
+      if (PN->getNumIncomingValues() == 2) {
+        // Recursive GEP in second incoming value. Always keep Recursive GEP as
+        // FirstInst
+        Value *FirstInst = nullptr, *SecondInst = nullptr;
+        for (unsigned i = 0; i != 2; ++i) {
+          // Check if LHS is a Recursive GEP in one of the incoming values.
+          if (PN->getIncomingValue(i) == LHS && LHSGEP->getNumIndices() == 1 &&
+              isa<Constant>(LHSGEP->idx_begin())) {
+            FirstInst = PN->getIncomingValue(i);
+            SecondInst = PN->getIncomingValue(!i);
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:2445]: `!i` -> `1-i` imo is clearer.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 4]
```diff
@@ -2410,9 +2410,76 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
...
+          FirstInst = FirstInst->stripAndAccumulateConstantOffsets(
+              Q.DL, FirstInstOffset, /* AllowNonInbounds */ true);
+          APInt SecondInstOffset(
+              Q.DL.getIndexTypeSizeInBits(SecondInst->getType()), 0);
+          SecondInst = SecondInst->stripAndAccumulateConstantOffsets(
+              Q.DL, SecondInstOffset, /* AllowNonInbounds */ true);
+          APInt RHSOffset(Q.DL.getIndexTypeSizeInBits(RHS->getType()), 0);
+          RHS = RHS->stripAndAccumulateConstantOffsets(
+              Q.DL, RHSOffset, /* AllowNonInbounds */ true);
+          if (SecondInst == RHS &&
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:2466]: Can you explain this check? It looks like its handling something like:
`(sub (ptrint (phi X, Y)), (ptrint y))` which very much so can be zero.

File: llvm/test/Analysis/ValueTracking/phi-known-bits.ll [hunk 5]
```diff
@@ -424,3 +424,330 @@ T:
...
+
+while.end.i:
+  %sub.ptr.lhs.cast.i = ptrtoint ptr %test.0.i to i64
+  %sub.ptr.rhs.cast.i = ptrtoint ptr %val1 to i64
+  %sub.ptr.sub.i = sub i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
+  br label %_Z9stringlenPKs.exit
+
+_Z9stringlenPKs.exit:
+  %retval.0.i = phi i64 [ %sub.ptr.sub.i, %while.end.i ], [ 0, %entry ]
+  %bool = icmp eq i64 %retval.0.i, 0
```
Comment (nikic) [inline llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: This code will get converted into something like this by InstCombine:
```suggestion
  %x = icmp eq i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
  br label %_Z9stringlenPKs.exit

_Z9stringlenPKs.exit:
  %bool = phi i1 [ %x, %while.end.i ], [ true, %entry ]
```
At which point the sub of ptrtoint pattern no longer exists.

So I don't understand how your code even does anything here -- unless you are relying on your fold triggering before this canonicalization, which is obviously not something you can do.

Comment (bipmis) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: Thanks for the review. 
For the given example code visitPHINode() should be invoked first and the code then identifies it as a knownNonZero() scenario and does the transformation. Exclusively this can be a check done in visitIcmp() too before we fold the icmp(sub) which could help fold additional scenarios.
Additionally since we are identifying it is a isNonZeroSub() this can be applied beyond scenarios where the only use of phi is a icmp. This can also be applied to 
1. Multiple uses of phi with or, icmp
2. Single use of phi with or(icmp)
Some of these examples can be added for the patch.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 4 above)
Comment (bipmis) [reply llvm/lib/Analysis/ValueTracking.cpp:2466]: Thanks for the review.
It is handling a recursive GEP as an operand to the sub, so something like
(sub (ptrint (gep(phi X, Y))), (ptrint y))
We handle 2 incoming Values to PHI, in this case one being the recursive GEP itself. So in this case if we can identify that the second incoming ptr Y>y and GEP offset is positive, the sub is always greater than zero. Similarly if Y<y and GEP offset is negative, the sub is always less than zero. We can build the isKnownNonZero based on this.
The scenario provided by you should not apply in this case. I can add an example for the same.

File: llvm/test/Analysis/ValueTracking/phi-known-bits.ll (hunk 5 above)
Comment (nikic) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: Okay, so you *are* relying on transform order. In that case, what you need to do is first implement the fold for canonical IR (that is, for the case where you have a `icmp eq i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i` style comparison). Once that has landed, we can go back and consider whether it is necessary or worthwhile to handle the non-canonical form.

Comment (bipmis) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: Ok so in this patch we are handling the sub of 2 ptr2ints.
But I am not specifically relying on the transform order for this pattern. Sure in the example provided the phi in InstCombine hits first. However there is use to the isNonZeroSub() from InstCombineCompares as well. Which means for the scneario where we have the below
```
  %sub.ptr.lhs.cast.i = ptrtoint ptr %test.0.i to i64
  %sub.ptr.rhs.cast.i = ptrtoint ptr %val1 to i64
  %sub.ptr.sub.i = sub i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
  %bool = icmp eq i64 %sub.ptr.sub.i, 0
  br label %_Z9stringlenPKs.exit

_Z9stringlenPKs.exit:
  %retval.0.i = phi i1 [ %bool, %while.end.i ], [ false, %entry ]
```

The same does get folded away as well. As I specified maybe adding some test cases showing this would give more clarity.
So basically irrespective of Transform order if we have a sub(pt2int, ptr2int) it should be folded away.

Comment (bipmis) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: Updated the PR with test showing folds with icmp(sub(ptr2int,ptr2int), 0) and scenarios where it wont fold from Compares(multi use phi, single use with or(icmp))

Comment (nikic) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: This doesn't really address my concern. My problem here is that you are trying to handle a complex pattern, while there exists a simpler pattern that is not being handled yet. You need to handle the simple pattern *first*, and the complex pattern later.

Comment (bipmis) [reply llvm/test/Analysis/ValueTracking/phi-known-bits.ll:463]: Thanks. That makes sense. I have done this in the PR https://github.com/llvm/llvm-project/pull/70459.

**Comment by bipmis** [comment]
@nikic Updated patch to handle the complex patterns.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 6]
```diff
@@ -2365,9 +2365,69 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
       .isNonZero();
 }
 
+// Check to see if A is both a GEP and is the incoming value for a PHI in the
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2368]: Can you please pre-commit the move of this function?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 7]
```diff
@@ -2365,9 +2365,69 @@ static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
...
+          (StartOffset.sle(OffsetB) && StepOffset.isNegative()));
+}
+
 static bool isNonZeroSub(const APInt &DemandedElts, unsigned Depth,
                          const SimplifyQuery &Q, unsigned BitWidth, Value *X,
                          Value *Y) {
+  // sub(PtrtoInt(A), PtrtoInt(B))
+  // Sub is a NonZero if the Ptr A and B are NonEqual.
+  Value *A, *B;
+  if (match(X, m_PtrToInt(m_Value(A))) && match(Y, m_PtrToInt(m_Value(B))))
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2426]: You need to check here that the ptrtoints are to the same type and that it also matches the pointer size. Otherwise you may have truncation. (Please test this, but note that it will require a `-passes=instsimplify` test, because `-passes=instcombine` will canonicalize it away.)

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 6 above)
Comment (nikic) [reply llvm/lib/Analysis/ValueTracking.cpp:2368]: Actually, I don't think we need this. Can you please rebase this on top of https://github.com/llvm/llvm-project/commit/337504683efa0b898bcf69b4e5be67d73dbaf180 and then add the ptrtoint handling inside isKnownNonEqual()? I think that is the cleanest way to handle this.

isKnownNonZero for sub will call isKnownNonEqual, that will look through ptrtoint and call itself recursively, and at that point we'll handle the GEP case. This will make sure that if we add any extra pointer cases, we'll automatically make use of them as well.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 7 above)
Comment (bipmis) [reply llvm/lib/Analysis/ValueTracking.cpp]: Thanks. Handled the same.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 6 above)
Comment (bipmis) [reply llvm/lib/Analysis/ValueTracking.cpp]: Agreed. This is a good improvement. Thanks for the suggestion.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 8]
```diff
@@ -3199,6 +3199,14 @@ static bool isKnownNonEqual(const Value *V1, const Value *V2, unsigned Depth,
       isNonEqualPointersWithRecursiveGEP(V2, V1, Q))
     return true;
 
+  Value *A, *B;
+  // PtrToInts are NonEqual if their Ptrs are NonEqual.
+  // Check PtrToInt type matches the pointer size.
+  if (match(V1, m_PtrToInt(m_Value(A))) && match(V2, m_PtrToInt(m_Value(B))) &&
+      Q.DL.getTypeSizeInBits(A->getType()) ==
+          Q.DL.getTypeSizeInBits(V1->getType()))
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:3207]:
```suggestion
  if (match(V1, m_PtrToIntSameSize(Q.DL, m_Value(A))) &&
      match(V2, m_PtrToIntSameSize(Q.DL, m_Value(B))))
```

**Review by nikic — APPROVED** [review]
LGTM
